pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+			Name: "alertmanager_partial_state_merges_total",
+			Help: "Number of times we have received a partial state to merge for a key.",
+		}, []string{"key"}),
+		partialStateMergesFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_failed_total",
+			Help: "Number of times we have failed to merge a partial state received for a key.",
+		}, []string{"key"}),
+		stateReplicationTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_total",
+			Help: "Number of times we have tried to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+		stateReplicationFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_failed_total",
+			Help: "Number of times we have failed to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+	}
+
+	go s.handleBroadcasts()
+	return s
+}
+
+// AddState adds a new state that will be replicated using the ReplicationFunc. It returns a channel to which the client can broadcast messages of the state to be sent.
+func (s *state) AddState(key string, cs cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return newStateChannel(key, send, s.logger)
+}
+
+// MergePartialState merges a received partial message with an internal state.
+func (s *state) MergePartialState(p *clusterpb.Part) error {
+	s.partialStateMergesTotal.WithLabelValues(p.Key).Inc()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	st, ok := s.states[p.Key]
+	if !ok {
+		level.Error(s.logger).Log("msg", "state key not found while merging", "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return fmt.Errorf("unable to merge state, key '%s' not found while merging", p.Key)
+	}
+
+	if err := st.Merge(p.Data); err != nil {
+		level.Error(s.logger).Log("msg", "failed to merge state", "err", err, "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return errors.Wrap(err, "unable to merge state")
+	}
+
+	return nil
+}
+
+// The clusterWait function uses the position to determine how long should it wait before sending a notification.
+// In the Cortex case, we copy the function directly and use the manager to figure out which position we hold.
+func (s *state) Position() int { return 0 }
+
+// Settle waits until the alertmanagers are ready (and sets the appropriate internal state when it is).
+// The idea is that we don't want to start working" before we get a chance to know most of the notifications and/or silences.
+func (s *state) Settle(ctx context.Context, _ time.Duration) {
```

Comment:
Settle ignores context currently. It should not do that, and should report error back if context is finished before settling has finished.

Reply:
I'm leaving the implementation of `WaitReady` and `Settle` to the next PR - is it OK If we skip it for now? These two involve a full state replication and I fear we need the implementation of those to make sense of the big picture.